engine/src/main/java/org/pentaho/di/www/DatabaseFeatureListServlet.java

```diff
@@ -0,0 +1,161 @@
+      DatabaseMeta databaseMeta = fetchDatabaseMeta( dbConnectionRequest );
+
+      List<Object[]> featureList = new LinkedList<>();
+      for ( RowMetaAndData rowMetaAndData : databaseMeta.getFeatureSummary() ) {
+        featureList.add( rowMetaAndData.getData() );
+      }
+
+      JSONArray featureListJson = new JSONArray();
+      for ( Object[] feature : featureList ) {
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put( feature[0] , feature[1] );
+        featureListJson.add( jsonObject );
+      }
+
+      JSONObject jsonOutput = new JSONObject();
+      jsonOutput.put( "featureList", featureListJson );
+
+      response.setStatus( HttpServletResponse.SC_OK );
+      out.print( jsonOutput );
+    } catch ( Exception ex ) {
+      response.setStatus( HttpServletResponse.SC_BAD_REQUEST );
+      out.print( ex.getMessage() );
+    }
+  }
+
+  @Override
+  public String getContextPath() {
+    return CONTEXT_PATH;
+  }
+
+  @Override
+  public String getService() {
+    return CONTEXT_PATH + " (" + this + ")";
+  }
+
+  private DatabaseMeta fetchDatabaseMeta( String dbConnectionRequest ) throws Exception {
+    JSONParser parser = new JSONParser();
+    try {
+      if( StringUtils.isBlank( dbConnectionRequest ) ) {
+        throw new Exception( "Invalid request for feature list" );
+      }
+
+      JSONObject dbConnectionJson = ( JSONObject ) parser.parse( dbConnectionRequest );
+
+      String name = ( String ) dbConnectionJson.get( "connectionName" );
+      String connectionType = ( String ) dbConnectionJson.get( "type" );
+      String accessType = ( String ) dbConnectionJson.get( "access" );
+      String hostName = ( String ) dbConnectionJson.get( "serverName" );
+      String database = ( String ) dbConnectionJson.get( "database" );
+      String port = ( String ) dbConnectionJson.get( "port" );
+      String username = ( String ) dbConnectionJson.get( "username" );
+      String password = ( String ) dbConnectionJson.get( "password" );
+
+      DatabaseMeta databaseMeta = new DatabaseMeta( name, connectionType, accessType, hostName,
+          database, port, username, password );
+
+      JSONArray attributesJson = ( JSONArray ) dbConnectionJson.get( "attributes" );
+      if( !Objects.isNull( attributesJson ) ) {
+        for ( Object object : attributesJson ) {
+          JSONObject jsonObject = ( JSONObject ) object;
+          String attribute = ( String ) jsonObject.get( "attribute" );
+          String value = ( String ) jsonObject.get( "code" );
+          validateAndAddAttribute( attribute, value, databaseMeta );
+        }
+      }
+
+      JSONArray parametersArray = ( JSONArray ) dbConnectionJson.get( "parameters" );
+      if( !Objects.isNull( parametersArray ) ) {
+        for ( Object object : parametersArray ) {
+          JSONObject jsonObject = ( JSONObject ) object;
+          String parameter = ( String ) jsonObject.get("parameter");
```

Comment:
Suggest to use String.valueOf(...)  to avoid `ClassCastException` (runtime/unchecked)

https://stackoverflow.com/questions/16815279/difference-between-casting-to-string-and-string-valueof#:~:text=Casting%20means%20that%20the%20object,take%20other%20types%20as%20well.&text=Both%20generates%20same%20output%20in,object%20is%20Not%20a%20string.